diff --git a/src/rp2_common/pico_cyw43_driver/cyw43_bus_pio_spi.c b/src/rp2_common/pico_cyw43_driver/cyw43_bus_pio_spi.c
index 5afe85e..75547ee 100644
--- a/src/rp2_common/pico_cyw43_driver/cyw43_bus_pio_spi.c
+++ b/src/rp2_common/pico_cyw43_driver/cyw43_bus_pio_spi.c
@@ -20,13 +20,13 @@
 #include "cyw43_debug_pins.h"
 
 #if CYW43_SPI_PIO
-#define WL_REG_ON 23
-#define DATA_OUT_PIN 24u
-#define DATA_IN_PIN 24u
-#define IRQ_PIN 24u
+#define WL_REG_ON 16u  /*23*/
+#define DATA_OUT_PIN 19u  /*24u*/
+#define DATA_IN_PIN 19u  /*24u*/
+#define IRQ_PIN 19u /*24u*/
 // #define MONITOR_PIN 3u
-#define CLOCK_PIN 29u
-#define CS_PIN 25u
+#define CLOCK_PIN 18u /*29u*/
+#define CS_PIN 17u /*25u*/
 #define IRQ_SAMPLE_DELAY_NS 100
 
 #define SPI_PROGRAM_NAME spi_gap01_sample0
